pdns/syncres.cc

```diff
@@ -4459,3 +4461,161 @@ int SyncRes::getRootNS(struct timeval now, asyncresolve_t asyncCallback, unsigne
+    }
+    catch (const PDNSException &e) {
+      g_log << Logger::Error << msg << e.reason << endl;
+    }
+    catch (const std::runtime_error &e) {
+      g_log << Logger::Error << msg << e.what() << endl;
+    }
+    catch (...) {
+      g_log << Logger::Error << msg << "unexpected exception" << endl;
+    }
+
+    // We do not want to refresh more than once per hour
+    refresh = max(refresh, 3600LL);
+  }
+
+  return refresh;
+}
+
+time_t SyncRes::ZoneToCache(const string& name, const string& url, bool dnssec)
+{
+  time_t refresh = std::numeric_limits<time_t>::max();
+
+#ifdef HAVE_LIBCURL
+
+  MiniCurl mc;
+  string reply = mc.getURL(url, nullptr, nullptr, 10, false, true);
+  vector<string> lines;
+  std::istringstream stream(reply);
+  string line;
+  while (std::getline(stream, line)) {
+    lines.push_back(line);
+  }
+
+  const DNSName zone(name);
+  DNSResourceRecord drr;
+  DNSName qname;
+  QType qtype;
+  vector<DNSRecord> v;
+  vector<DNSRecord> sigs;
+  vector<shared_ptr<RRSIGRecordContent>> sigsrr;
+  bool first = true;
+  time_t now = time(nullptr);
+  recordsAndSignatures soa;
+  uint32_t last_ttl = 0;
+
+  const string msg  = "zones-to-cache error while loading " + name + " from " + url + ": ";
+  ZoneParserTNG zpt(lines, zone);
+  zpt.setMaxGenerateSteps(0);
+
+  // The code below assumes the records for a name,type tuple are grouped together and followed by their corresponding sigs.
+  // We do not do validation, it will happen on-demand if an Indeterminate record is encountered when the caches are queried
+  while (zpt.get(drr)) {
+    DNSRecord dr(drr);
+    if (first) {
+      qname = dr.d_name;
+      qtype = dr.d_type;
+      first = false;
+    }
+    if (qname != dr.d_name || qtype != dr.d_type) {
+      // We came to a different block
+      if (qname == dr.d_name && dr.d_type == QType::RRSIG) {
+        // same name,, but now collect sigs for it
+        qtype = QType::RRSIG;
+      } else {
+        // done with the previous block of records and corresponding sigs, store them into the caches
+        if (qtype == QType::SOA) {
+          if (soa.records.size() > 0) {
+            g_log << Logger::Error << msg << "can only load zones with a single SOA record ";
+            return refresh;
+          }
+          soa.records = v;
+          if (dnssec) {
+            soa.signatures = sigs;
+          }
+        }
+        switch (qtype) {
+        case QType::NSEC:
+        case QType::NSEC3: {
+          if (soa.records.size() == 0) {
+            g_log << Logger::Error << msg << "SOA record should precede any NSEC/NSEC3 records";
+            return refresh;
+          }
+          NegCache::NegCacheEntry ne;
+          ne.authoritySOA = soa;
+          ne.DNSSECRecords.records = v;
+          ne.DNSSECRecords.signatures = sigs;
+          ne.d_name = qname;
+          ne.d_auth = zone;
+          ne.d_ttd = now + last_ttl;
+          ne.d_qtype = qtype;
+          g_negCache->add(ne);
```

Comment:
I don't understand what we are negatively caching? At this point `qname` and `qtype` are still from the previous entry, and having a `NSEC` for a given `qname` is **not** a denial of that name. It's even more complicated for a `NSEC3`, of course.

Reply:
I should have left out this part. It was a misguided attempt.